include/extractor/guidance/graph_hopper.hpp

```diff
@@ -0,0 +1,143 @@
+                                                            accumulator_type &accumulator,
+                                                            const selector_type &selector);
+
+  private:
+    const util::NodeBasedDynamicGraph &node_based_graph;
+    const IntersectionGenerator &intersection_generator;
+};
+
+// Accumulate all coordinates following a road until we
+struct LengthLimitedCoordinateAccumulator
+{
+    LengthLimitedCoordinateAccumulator(
+        const extractor::guidance::CoordinateExtractor &coordinate_extractor,
+        const double max_length);
+
+    // true if the path has traversed enough distance
+    bool terminate();
+
+    // update the accumulator
+    void update(const NodeID from_node,
+                const EdgeID via_edge,
+                const NodeID to_node,
+                const util::NodeBasedEdgeData &edge_data);
+
+    const extractor::guidance::CoordinateExtractor &coordinate_extractor;
+    const double max_length;
+    double accumulated_length;
+    std::vector<util::Coordinate> coordinates;
+};
+
+// The FollowRoadNameSelector tries to follow a given name along a route. We offer methods to skip
+// over bridges/similar situations if desired, following narrow turns
+struct SelectRoadByNameOnlyChoiceAndStraightness
+{
+    SelectRoadByNameOnlyChoiceAndStraightness(const NameID desired_name_id,
+                                              const bool requires_entry);
+
+    // we expect the node we are coming form, the edge we reach the intersection over, the
+    // intersection itself as well as the graph to
+    boost::optional<EdgeID> operator()(const NodeID nid,
+                                       const EdgeID via_edge_id,
+                                       const Intersection &intersection,
+                                       const util::NodeBasedDynamicGraph &node_based_graph) const;
+
+    const NameID desired_name_id;
+    const bool requires_entry;
+};
+
+template <class accumulator_type, class selector_type>
+boost::optional<std::pair<NodeID, EdgeID>> GraphHopper::TraverseRoad(NodeID current_node_id,
+                                                                     EdgeID current_edge_id,
+                                                                     accumulator_type &accumulator,
+                                                                     const selector_type &selector)
+{
+    // since graph hopping is used in many ways, we don't generate an adjusted intersection
+    // (otherwise we could end up in infinite recursion if we call the graph hopper during the
+    // adjustment itself). Relying only on `GetConnectedRoads` (which itself does no graph hopping),
+    // we prevent this from happening.
+    const auto stop_node_id = current_node_id;
+    // we wan't to put out the last valid entries. To do so, we need to update within the following
+    // while loop.
+    for (std::size_t safety_hop_limit = 0; safety_hop_limit < 1000; ++safety_hop_limit)
```

Comment:
if you look upwards, we are also doing this here (stop_node_id). I just wanted to ensure we don't get caught in endless loops again (due to some random graph issues) and prevent any while loops in the code. If the graph traversal should ever be looking at more than 10 intersections, we are probably doing something wrong anyhow. Having this limit in here simply assures that we will get notice but the graph will not simply hang forever.